File: aramis-opengl/src/main/java/hello/com/aramis/opengl/december/utils/CameraHelper.java
package hello.com.aramis.opengl.december.utils;

import android.graphics.ImageFormat;
import android.graphics.SurfaceTexture;
import android.hardware.Camera;

/**
 * Created by Aramis
 * Date:2018/12/5
 * Description:
 */
public class CameraHelper implements Camera.PreviewCallback {

    private static final String TAG = "CameraHelper";
    public static final int WIDTH = 1280;
    public static final int HEIGHT = 720;
    private int mCameraId;
    private Camera mCamera;
    private byte[] buffer;
    private Camera.PreviewCallback mPreviewCallback;
    private SurfaceTexture mSurfaceTexture;

    public CameraHelper(int cameraId) {
        mCameraId = cameraId;
    }

    public void switchCamera() {
        if (mCameraId == Camera.CameraInfo.CAMERA_FACING_BACK) {
            mCameraId = Camera.CameraInfo.CAMERA_FACING_FRONT;
        } else {
            mCameraId = Camera.CameraInfo.CAMERA_FACING_BACK;
        }
        stopPreview();
        startPreview(mSurfaceTexture);
    }

    public int getCameraId() {
        return mCameraId;
    }

    public void stopPreview() {
        if (mCamera != null) {
            //预览数据回调接口
            mCamera.setPreviewCallback(null);
            //停止预览
            mCamera.stopPreview();
            //释放摄像头
            mCamera.release();
            mCamera = null;
        }
    }

    private Camera.Parameters mCameraParameters;

    public void startPreview(SurfaceTexture surfaceTexture) {
        mSurfaceTexture = surfaceTexture;
        try {
            //获得camera对象
            mCamera = Camera.open(mCameraId);
            //配置camera的属性
             mCameraParameters = mCamera.getParameters();
            //设置预览数据格式为nv21
            mCameraParameters.setPreviewFormat(ImageFormat.NV21);
            //这是摄像头宽、高
            mCameraParameters.setPreviewSize(WIDTH, HEIGHT);
            // 设置摄像头 图像传感器的角度、方向
            mCamera.setParameters(mCameraParameters);
            buffer = new byte[WIDTH * HEIGHT * 3 / 2];
            //数据缓存区
            mCamera.addCallbackBuffer(buffer);
            mCamera.setPreviewCallbackWithBuffer(this);
            //设置预览画面
            mCamera.setPreviewTexture(mSurfaceTexture);
            mCamera.startPreview();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }


    public void setPreviewCallback(Camera.PreviewCallback previewCallback) {
        mPreviewCallback = previewCallback;
    }


    @Override
    public void onPreviewFrame(byte[] data, Camera camera) {
        // data数据依然是倒的
        if (null != mPreviewCallback) {
            mPreviewCallback.onPreviewFrame(data, camera);
        }
        camera.addCallbackBuffer(buffer);
    }


    public void autoFocus(){
        if (mCamera!=null) {
            mCamera.autoFocus(new Camera.AutoFocusCallback() {

                @Override
                public void onAutoFocus(boolean success, Camera camera) {
                    if (success) {
                        System.out.println(">>>>>>>>success");
                    }else{
                        camera.autoFocus(this);//如果失败，自动聚焦
                    }
                }
            });
        }

    }


}

File: aramis-opengl/src/main/java/hello/com/aramis/opengl/douyin/utils/CameraHelper.kt
package hello.com.aramis.opengl.douyin.utils

import android.graphics.ImageFormat
import android.graphics.SurfaceTexture
import android.hardware.Camera
import java.lang.Exception


/**
 *Created by Aramis
 *Date:2018/10/31
 *Description:
 */
class CameraHelper(var cameraID: Int) : android.hardware.Camera.PreviewCallback {
    private val TAG = "CameraHelper"
    var WIDTH = 640
    var HEIGHT = 480
    private var mCamera: Camera? = null
    private lateinit var buffer: ByteArray
    var mPreviewCallback: Camera.PreviewCallback? = null
    private lateinit var mSurfaceTexture: SurfaceTexture

    fun switchCamera() {
        cameraID = if (cameraID == Camera.CameraInfo.CAMERA_FACING_BACK) {
            Camera.CameraInfo.CAMERA_FACING_FRONT
        } else {
            Camera.CameraInfo.CAMERA_FACING_BACK
        }
        stopPreview()
        startPreview(mSurfaceTexture)
    }

    fun startPreview(surfaceTexture: SurfaceTexture) {
        mSurfaceTexture = surfaceTexture
        try {
            mCamera = Camera.open(cameraID)
            mCamera?.apply {
                //配置Camera属性
                val parameters = this.parameters
                //设置预览数据格式为nv21
                parameters.previewFormat = ImageFormat.NV21
                //设置摄像头的宽高
                parameters.setPreviewSize(WIDTH, HEIGHT)
                //
                this.parameters = parameters
                buffer = ByteArray(WIDTH * HEIGHT * 3 / 2)
                this.addCallbackBuffer(buffer)
                this.setPreviewCallbackWithBuffer(this@CameraHelper)
                //设置预览画面
                this.setPreviewTexture(mSurfaceTexture)
                this.startPreview()
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun stopPreview() {
        mCamera?.apply {
            //移除回调接口
            this.setPreviewCallback(null)
            //停止预览
            this.stopPreview()
            //释放摄像头
            this.release()
            mCamera = null
        }
    }

    override fun onPreviewFrame(data: ByteArray?, camera: Camera?) {
        //data数据依然是倒的
        if (mPreviewCallback != null) {
            mPreviewCallback?.onPreviewFrame(data, camera)
        }
        camera?.addCallbackBuffer(buffer)
    }
}